File: 155-min-stack/min-stack.cpp
class MinStack {
    stack<long> st;
    long minElement;
public:
    MinStack() {
        minElement = INT_MAX;
    }
    
    void push(long val) {
        if (st.empty()) {
            st.push(val);
            minElement = val;
        } else {
            if (val < minElement) {
                long diff = 2 * val - minElement;
                st.push(diff);
                minElement = val;
            } else {
                st.push(val);
            }
        }
    }
    
    void pop() {
        long diff = st.top();
        if (diff < minElement) {
            minElement = 2 * minElement - diff;
        }
        st.pop();        
    }
    
    int top() {
        long diff = st.top();
        if (diff < minElement) {
            return minElement;
        }
        return diff;
    }
    
    int getMin() {
        return minElement;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * long param_3 = obj->top();
 * long param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.java
class MinStack {
    private Stack<Long> st = null;
    private long minElement;
    public MinStack() {
        st = new Stack<>();
        minElement = Long.MAX_VALUE;
    }
    
    public void push(int val) {
        if (st.isEmpty()) {
            st.push((long)val);
            minElement = val;
        } else {
            if (val < minElement) {
                long diff = 2L * val - minElement;
                st.push(diff);
                minElement = val;
            } else {
                st.push((long)val);
            }
        }
    }
    
    public void pop() {
        long diff = st.peek();
        if (diff < minElement) {
            minElement = 2L * minElement - diff;
        }
        st.pop();
    }
    
    public int top() {
        long diff = st.peek();
        if (diff < minElement) {
            return (int)minElement;
        }
        return (int)diff;
    }
    
    public int getMin() {
        return (int)minElement;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */